### PR TITLE
Added operation context to authentication

### PR DESCRIPTION
**Pull request by rozza**
JAVA-5211

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/connection/CommandHelper.java [hunk 1]
```diff
@@ -47,24 +47,23 @@ public final class CommandHelper {
...
         } catch (MongoServerException e) {
             return new BsonDocument();
         }
     }
 
     static void executeCommandAsync(final String database, final BsonDocument command, final ClusterConnectionMode clusterConnectionMode,
-                                    @Nullable final ServerApi serverApi, final InternalConnection internalConnection,
-                                    final SingleResultCallback<BsonDocument> callback) {
+            @Nullable final ServerApi serverApi, final InternalConnection internalConnection, final OperationContext operationContext,
+            final SingleResultCallback<BsonDocument> callback) {
```
Comment (vbabanin) [inline driver-core/src/main/com/mongodb/internal/connection/CommandHelper.java:66]: `OperationContext` isn't being used for socket communications here. It seems our unified tests might not be covering the socket interaction thoroughly. Should we consider adding some unit tests to ensure we have comprehensive coverage in this area?

Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/connection/CommandHelper.java:66]: Great catch - thanks.

Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/connection/CommandHelper.java:62]: Done

File: driver-core/src/main/com/mongodb/internal/async/function/RetryState.java [hunk 2]
```diff
@@ -182,6 +182,9 @@ private void doAdvanceOrThrow(final Throwable attemptException,
             final boolean onlyRuntimeExceptions) throws Throwable {
         assertTrue(attempt() < attempts);
         assertNotNull(attemptException);
+        if (attemptException instanceof MongoOperationTimeoutException) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/async/function/RetryState.java:185]: This conditional seems unrelated to the rest of the PR. What's the intention of including it?

Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/async/function/RetryState.java:185]: Had a test failing due to the wrapping of the exception that later occurs.

File: driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideOperationsTimeoutProseTest.java [hunk 3]
```diff
@@ -337,6 +461,12 @@ static Stream<Arguments> test8ServerSelectionArguments() {
         );
     }
 
+    private static Stream<Arguments> test8ServerSelectionHandshakeArguments() {
+        return Stream.of(
+                Arguments.of("timeoutMS honored for connection handshake commands if it's lower than serverSelectionTimeoutMS", 100, 200),
```
Comment (vbabanin) [inline driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideOperationsTimeoutProseTest.java:466]: In this test scenario, increasing `timeoutMS `to 500ms prevents` MongoTimeoutException` from occurring. However, as I see it, it's expected to be thrown regardless of the timeoutMS value.

The spec suggests using the minimum of `remaining computedServerSelectionTimeout` and `timeoutMS `for authentication. Currently, authentication relies solely on `timeoutMS`, ignoring `serverSelectionTimeout`.

Considering [DRIVERS-2750](https://jira.mongodb.org/browse/DRIVERS-2750), which proposes using timeoutMS exclusively for connection checkout, I'm curious if these changes align with the spec or the [ticket](https://jira.mongodb.org/browse/DRIVERS-2750). If it's the latter, perhaps a custom test case should verify this behavior, as the current spec test doesn't match the expected behavior.

If it's the former, we should ensure remaining `serverSelectionTimeout` is considered in authentication.

File: driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideOperationsTimeoutProseTest.java [hunk 4]
```diff
@@ -337,6 +461,12 @@ static Stream<Arguments> test8ServerSelectionArguments() {
...
+                Arguments.of("serverSelectionTimeoutMS honored for connection handshake commands if it's lower than timeoutMS", 200, 100)
```
Comment (vbabanin) [inline driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideOperationsTimeoutProseTest.java:467]: In this test scenario, increasing timeoutMS to 500ms prevents MongoTimeoutException from occurring. However, as I see it, it's expected to be thrown regardless of the timeoutMS value.

The spec suggests using the minimum of remaining computedServerSelectionTimeout and timeoutMS for authentication. Currently, authentication relies solely on timeoutMS, ignoring serverSelectionTimeout.

Considering [DRIVERS-2750](https://jira.mongodb.org/browse/DRIVERS-2750), which proposes using timeoutMS exclusively for connection checkout, I'm curious if these changes align with the spec or the [ticket](https://jira.mongodb.org/browse/DRIVERS-2750). If it's the latter, perhaps a custom test case should verify this behavior, as the current spec test doesn't match the expected behavior.

If it's the former, we should ensure remaining serverSelectionTimeout is considered in authentication.

Comment (vbabanin) [inline driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideOperationsTimeoutProseTest.java:467]: In this test scenario, increasing `timeoutMS` to 500ms prevents `MongoTimeoutException` from occurring. However, as I see it, it's expected to be thrown regardless of the `timeoutMS` value.

The spec suggests using the minimum of `remaining computedServerSelectionTimeout` and `timeoutMS` for authentication. Currently, authentication relies solely on `timeoutMS`, ignoring `remaining serverSelectionTimeou`t.

Considering [DRIVERS-2750](https://jira.mongodb.org/browse/DRIVERS-2750), which proposes using `timeoutMS` for connection checkout, I'm curious if these changes align with the spec or the [ticket](https://jira.mongodb.org/browse/DRIVERS-2750). If it's the latter, perhaps a custom test case should verify this behavior, as the current spec test doesn't match the expected behavior.

If it's the former, we should ensure `remaining serverSelectionTimeout` is considered in authentication.

**Review by vbabanin — COMMENTED** [review]
Looks good! Just one question about the spec implementation.

File: driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideOperationsTimeoutProseTest.java (hunk 4 above)
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideOperationsTimeoutProseTest.java:467]: Agreed and that work is deferred to JAVA-5211 

From an authentication / handshake perspective there should be no change as it will continue to use the passed down TimeoutContext.

**Review by vbabanin — APPROVED** [review]
